addons/stock/models/stock_location.py

```diff
@@ -156,23 +158,44 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
-        # Looking for a putaway about the product.
-        putaway_rules = self.putaway_rule_ids.filtered(lambda x: x.product_id == product)
-        if putaway_rules:
-            putaway_location = putaway_rules[0].location_out_id
-        # If not product putaway found, we're looking with category so.
-        else:
-            categ = product.categ_id
-            while categ:
-                putaway_rules = self.putaway_rule_ids.filtered(lambda x: x.category_id == categ)
-                if putaway_rules:
-                    putaway_location = putaway_rules[0].location_out_id
-                    break
-                categ = categ.parent_id
-        return putaway_location
+    def _get_putaway_strategy(self, product, quantity=0, package=None):
+        """Returns the location where the product has to be put, if any compliant
+        putaway strategy is found. Otherwise returns self.
+        The quantity should be in the default UOM of the product, it is used when
+        no package is specified.
+        """
+        package_type = package and package.package_type_id or self.env['stock.package.type']
+
+        putaway_rules = self.env['stock.putaway.rule']
+        putaway_rules |= self.putaway_rule_ids.filtered(lambda x: x.product_id == product and (package_type in x.package_type_ids or package_type == x.package_type_ids))
+        categ = product.categ_id
+        while categ:
+            putaway_rules |= self.putaway_rule_ids.filtered(lambda x: x.category_id == categ and (package_type in x.package_type_ids or package_type == x.package_type_ids))
+            categ = categ.parent_id
+        if package_type:
+            putaway_rules |= self.putaway_rule_ids.filtered(lambda x: not x.product_id and (package_type in x.package_type_ids or package_type == x.package_type_ids))
+
+        qty_by_location = {}
+        if putaway_rules.storage_category_id:
+            # get current product qty (qty in current quants and future qty on assigned ml) of all child locations
+            locations = self.env['stock.location'].search([('id', 'child_of', self.id), ('usage', '=', 'internal')])
+            move_line_data = self.env['stock.move.line'].read_group([
+                ('product_id', '=', product.id),
+                ('location_dest_id', 'in', locations.ids),
+                ('state', 'not in', ['draft', 'done', 'cancel'])
+            ], ['location_dest_id', 'product_qty:sum'], ['location_dest_id'])
+            quant_data = self.env['stock.quant'].read_group([
+                ('product_id', '=', product.id),
+                ('location_id', 'in', locations.ids),
+            ], ['location_id', 'quantity:sum'], ['location_id'])
+            for values in move_line_data:
+                qty_by_location[values['location_dest_id'][0]] = values['product_qty']
+            for values in quant_data:
+                qty_by_location[values['location_id'][0]] = qty_by_location.get(values['location_id'][0], 0) + values['quantity']
```

Comment:
`qty_by_location[values['location_id'][0]] += values['quantity']` is more concise

Reply:
keyerror when no `qty_by_location[values['location_id'][0]]`?

Reply:
ah yep. I though you used a default dict instead of regular dict.